### ISSUE TITLE
Move lean({ virtuals: true }) support from mongoose-lean-virtuals into core Mongoose

### ISSUE THREAD

**Issue by Jokero** [Automattic/mongoose#12684]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.7.2

### Node.js version

18.12.0

### MongoDB server version

N/A

### Typescript version (if applicable)

4.8.4

### Description

`InferSchemaType` does not infer types for virtual fields when schema has any. According to this PR https://github.com/Automattic/mongoose/pull/11908, it should work.

### Steps to Reproduce

1) Copy example from mentioned PR https://github.com/Automattic/mongoose/pull/11908/files#diff-fc5a8155d58dc5904b33fcc871f11ee08cb6f6e7adbe69e7bfbd7d173e82d8d2 (`Automatically Inferred Types`), create a type with `InferSchemaType` and try to access `fullName`:
```ts
import { Schema, InferSchemaType } from 'mongoose';

const schema = new Schema(
    {
        firstName: String,
        lastName: String,
    },
    {
        virtuals:{
            fullName:{
                get(){
                    return `${this.firstName} ${this.lastName}`;
                }
                // virtual setter and options can be defined here as well.
            }
        }
    }
);

type User = InferSchemaType<typeof schema>;

const user: User = {};
user.fullName
```
2) You will get the following error: 
```
TS2339: Property 'fullName' does not exist on type '{ firstName?: string | undefined; lastName?: string | undefined; }'.
```

### Expected Behavior

Type `User` should contain `fullName: string` field, and `user.fullName` shouldn't fail TS compilation.

**Comment by carlosingles** [comment]
As a work around, you can use the `ObtainSchemaGeneric` type to add the virtual types:

```ts
import { Schema, InferSchemaType, ObtainSchemaGeneric } from 'mongoose';
...
type User = InferSchemaType<typeof schema> & ObtainSchemaGeneric<typeof schema, 'TVirtuals'>;

const user = {} as User;
user.fullName // works
```

**Comment by Jokero** [comment]
@carlosingles, thanks for the workaround, currently I hardcoded all virtuals in an interface. 

@vkarpov15, I've tested the example from my initial message in 6.7.4, and the issue still persists (`fullName` is not available in a type returned by`InferSchemaType`).

If speak generally, I think it makes sense to have a way to get both interfaces (with and without virtuals). Based on first message example:

1) full interface with virtuals which can be used with find queries
```ts
type FullUser = InferFullSchemaType<typeof schema>;

//{
//  _id: ObjectId;
//  id: string;
//  firstName: string;
//  lastName: string;
//  fullName: string;
//}

function findOneUser(): Promise<FullUser> {
  return UserModel.findOne().orFail().exec();
}
```
2) persistent interface (only with persisted fields and without _id) that can be used with creation method
```ts
type PersistentUserData = InferPersistentSchemaType<typeof schema>;

//{
//  firstName: string;
//  lastName: string;
//}

function createUser(data: PersistentUserData): Promise<FullUser> {
  return UserModel.create(data);
}
```
Theoretically `_id` field can be omitted manually, but this won't work if there are nested fields with `_id`.

**Comment by vkarpov15** [comment]
@Jokero I took a look and that's expected behavior. `InferSchemaType` returns the raw document interface, which represents how the data is stored in MongoDB. The Model does correctly contain the virtuals, try the following:

```ts
const UserModel = model('AutoTypedVirtuals', schema);
type User = ReturnType<(typeof UserModel)['hydrate']>;

const user: User = new UserModel({ firstName: 'foo' });
user.fullName;

```

**Comment by vkarpov15** [comment]
@Jokero are you using mongoose-lean-virtuals? `lean()` doesn't include virtuals by default.

**Comment by Jokero** [comment]
@vkarpov15 yes, I'm using it like this `lean({ autopopulate: true, virtuals: true })`. In some cases I also specify the type explicitly `lean<Quest | null>({ autopopulate: true, virtuals: true })`

**Comment by vkarpov15** [comment]
@Jokero we added a helper type and some TypeScript docs to mongoose-lean-virtuals here: https://github.com/mongoosejs/mongoose-lean-virtuals?tab=readme-ov-file#typescript . That should help a little bit. Realistically any type improvements related to this issue belong in mongoose-lean-virtuals, because without mongoose-lean-virtuals then `lean({ virtuals: true })` does nothing.

Will keep this issue open because it does bring up a good point: we may want to consider moving mongoose-lean-virtuals support into Mongoose core for better TypeScript support.
